state/execution.go

```diff
@@ -417,15 +416,44 @@ func (blockExec *BlockExecutor) Commit(
 	)
 
 	// Update mempool.
+	ch <- mempoolUpdate{block: block, state: state.Copy(), abciResponse: abciResponse, err: err}
+
+	return res.RetainHeight, nil
+}
+
+// updates the mempool with the latest state asynchronously.
+// it begins flushing the mempool immediately, and then updates the mempool
+// after the new block state is ready.
+func (blockExec *BlockExecutor) asyncUpdateMempool(ch chan mempoolUpdate) {
+	blockExec.mempool.Lock()
+	defer blockExec.mempool.Unlock()
+
+	// while mempool is Locked, flush to ensure all async requests have completed
+	// in the ABCI app before Commit.
+	err := blockExec.mempool.FlushAppConn()
+	if err != nil {
+		blockExec.logger.Error("client error during mempool.FlushAppConn, flushing mempool", "err", err)
+		blockExec.mempool.Flush()
+		return
+	}
+
+	updateData := <-ch
+	if updateData.err != nil {
+		return
+	}
+
 	err = blockExec.mempool.Update(
```

Comment:
This part we can do in parallel.

Once the new state is committed, the application can start performing the re-check of all transactions remaining in the mempool, using the updated state.

Reply:
Moreover, it is not clear to me why we need to call `Update` with a locked mempool...

Reply:
As far as I understand, this is due to the current FIFO guarantee, before you add a new tx. (You want to know your done rechecking, before you add more txs to the end)

This really should be a lock owned within the mempool though, rather than here.